kymatio/scattering2d/tests/test_numpy_scattering2d.py

```diff
@@ -0,0 +1,313 @@
+            backend.cdgmm(np.empty((3, 4, 5)).astype(np.complex128),
+                          np.empty((4, 6)).astype(np.complex128))
+        assert 'not compatible for multiplication' in record.value.args[0]
+
+        with pytest.raises(TypeError) as record:
+            backend.cdgmm(np.empty((3, 4, 5)).astype(np.complex128),
+                          np.empty((4, 5)).astype(np.complex64))
+        assert 'same base dtype' in record.value.args[0]
+
+
+class TestFFT:
+    @pytest.mark.parametrize('backend', backends)
+    def test_fft(self, backend):
+        x = np.random.randn(2, 2) + 1J * np.random.randn(2, 2)
+
+        y = np.array([[x[0, 0] + x[0, 1] + x[1, 0] + x[1, 1],
+                       x[0, 0] - x[0, 1] + x[1, 0] - x[1, 1]],
+                      [x[0, 0] + x[0, 1] - x[1, 0] - x[1, 1],
+                       x[0, 0] - x[0, 1] - x[1, 0] + x[1, 1]]])
+
+        z = backend.fft(x, direction='C2C')
+
+        assert np.allclose(y, z)
+
+        z = backend.fft(x, direction='C2C', inverse=True)
+
+        assert np.allclose(y, z)
+
+        z = backend.fft(x, direction='C2R', inverse=True)
+
+        assert not np.iscomplexobj(z)
+        assert np.allclose(np.real(y), z)
+
+
+    @pytest.mark.parametrize('backend', backends)
+    def test_fft_exceptions(self, backend):
+        with pytest.raises(RuntimeError) as record:
+            backend.fft(np.empty((2, 2)), direction='C2R',
+                        inverse=False)
+        assert 'done with an inverse' in record.value.args[0]
+
+
+class TestBackendUtils:
+    @pytest.mark.parametrize('backend', backends)
+    def test_empty_like(self, backend):
+        x = np.random.randn(2, 2) + 1J * np.random.randn(2, 2)
+        y = backend.empty_like(x, (3, 3))
+
+        assert y.shape == (3, 3)
+        assert y.dtype == x.dtype
+
+
+class TestScattering2DNumpy:
+    def reorder_coefficients_from_interleaved(self, J, L):
+        # helper function to obtain positions of order0, order1, order2 from interleaved
+        order0, order1, order2 = [], [], []
+        n_order0, n_order1, n_order2 = 1, J * L, L ** 2 * J * (J - 1) // 2
+        n = 0
+        order0.append(n)
+        for j1 in range(J):
+            for l1 in range(L):
+                n += 1
+                order1.append(n)
+                for j2 in range(j1 + 1, J):
+                    for l2 in range(L):
+                        n += 1
+                        order2.append(n)
+        assert len(order0) == n_order0
+        assert len(order1) == n_order1
+        assert len(order2) == n_order2
+        return order0, order1, order2
+
+
+    @pytest.mark.parametrize('backend', backends)
+    def test_Scattering2D(self, backend):
+        test_data_dir = os.path.dirname(__file__)
+        data = None
+        with open(os.path.join(test_data_dir, 'test_data_2d.pt'), 'rb') as f:
+            buffer = io.BytesIO(f.read())
+            data = torch.load(buffer)
+
+        x = data['x'].numpy().astype(np.zeros(1).dtype)
```

Comment:
hint

Reply:
Don't like this. Let's fix it once and for all.

Reply:
what if an user needs to change the type?

Reply:
See above.

Reply:
agreed